sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+            ON {self.table_identifier(dataset)} ({self.quote(geom_col)});
+            """
+        )
+        L.info("Created spatial index in %ss", time.monotonic() - t0)
+
+    def _drop_spatial_index(self, sess, dataset):
+        # SQL server deletes the spatial index automatically when the table is deleted.
+        pass
+
+    def _quoted_trigger_name(self, dataset):
+        trigger_name = f"{dataset.table_name}_sno_track"
+        return f"{self.DB_SCHEMA}.{self.quote(trigger_name)}"
+
+    def _create_triggers(self, sess, dataset):
+        pk_name = dataset.primary_key
+        escaped_table_name = dataset.table_name.replace("'", "''")
+
+        sess.execute(
+            f"""
+            CREATE TRIGGER {self._quoted_trigger_name(dataset)} ON {self.table_identifier(dataset)}
+            AFTER INSERT, UPDATE, DELETE AS
+            BEGIN
+                MERGE {self.SNO_TRACK} TRA
+                USING
+                    (SELECT '{escaped_table_name}', {self.quote(pk_name)} FROM inserted
+                    UNION SELECT '{escaped_table_name}', {self.quote(pk_name)} FROM deleted)
+                    AS SRC (table_name, pk)
+                ON SRC.table_name = TRA.table_name AND SRC.pk = TRA.pk
+                WHEN NOT MATCHED THEN INSERT (table_name, pk) VALUES (SRC.table_name, SRC.pk);
+            END;
+            """,
+            {"table_name": dataset.table_name},
+        )
+
+    @contextlib.contextmanager
+    def _suspend_triggers(self, sess, dataset):
+        sess.execute(
+            f"""DISABLE TRIGGER {self._quoted_trigger_name(dataset)} ON {self.table_identifier(dataset)};"""
+        )
+        yield
+        sess.execute(
+            f"""ENABLE TRIGGER {self._quoted_trigger_name(dataset)} ON {self.table_identifier(dataset)};"""
+        )
+
+    def meta_items(self, dataset):
+        with self.session() as sess:
+            table_info_sql = """
+                SELECT
+                    C.column_name, C.ordinal_position, C.data_type, C.udt_name,
+                    C.character_maximum_length, C.numeric_precision, C.numeric_scale,
+                    KCU.ordinal_position AS pk_ordinal_position,
+                    upper(postgis_typmod_type(A.atttypmod)) AS geometry_type,
+                    postgis_typmod_srid(A.atttypmod) AS geometry_srid
+                FROM information_schema.columns C
+                LEFT OUTER JOIN information_schema.key_column_usage KCU
+                ON (KCU.table_schema = C.table_schema)
+                AND (KCU.table_name = C.table_name)
+                AND (KCU.column_name = C.column_name)
+                LEFT OUTER JOIN pg_attribute A
+                ON (A.attname = C.column_name)
+                AND (A.attrelid = (C.table_schema || '.' || C.table_name)::regclass::oid)
+                WHERE C.table_schema=:table_schema AND C.table_name=:table_name
+                ORDER BY C.ordinal_position;
+            """
+            table_info_sql = """
```

Comment:
redefining the above variable...

Reply:
Oops. Removed first unneeded definition.

Reply:
Oops! Removed useless first definition